docs/checkedc/Implementation-Notes.md

```diff
@@ -21,7 +21,120 @@ to handle checked extensions at this time. LLVM/clang will only allow
 the Checked C extension to be used with C and will not allow it to
 be used with C++, Objective C, or OpenCL.
 
-## Processing extensions to the IR
+## IR Extensions
+
+### Types
+
+Pointer types and array types are extended with information about whether the
+types are checked.  For pointer types, an enum is used to represent the
+different kinds of pointers.  For array types, a boolean flag is used to
+represent whether the array type is checked or not.  This information is used
+when determining type compatibility: two pointers types are compatible only
+when their kinds are the same and two array types are compatible only when
+they have the same checked property.
+
+Function types are extended with optional bounds expressions for parameters
+and the return value.  For parameters, the bounds expressions are recorded
+in a variably-sized array allocated within the function type.  The array
+is not allocated if there are no bounds expressions on parameters.
+
+Bounds expressions are canonicalized for function types. References to
+parameters are abstracted to the index of the parameter in the argument
+list, so that names of arguments do not matter.
+
+### Bounds expressions
+
+Bounds expressions are represented using subclasses of the Expr
+class.  This allows the usual machinery for expressions to be
+reused.  There are three kinds of bounds expressions:
+
+- Nullary bounds expressions.  These do not refer to any other
+  expressions.  There are two kinds of nullary bounds expressions:
+  None and Invalid.  Invalid represents bounds expressions that
+  are invalid for some reason.  They are used to prevent errors
+  from cascading during semantic processing.  Just ignoring an invalid
+  bounds expressions could trigger downstream semantic checking errors.
+- Count bounds expressions, which refer to a single expression
+  that is the count.  There are two kinds of count bounds expressions:
+  element count and byte counts.
+- Range bounds expressions, which refer to lower and upper bound
+  expressions.
+
+There is an abtract base class `BoundsExpr` for bounds expressions.
+A unified kind is stored on this base class.
+
+Type annotations are also represented as a subclass of `BoundsExpr`.
+They can appear syntactically where bounds expressions can apppear.
+The type for the type annotation is stored as the type of the
+type annotation expression.
+
+### Other expressions
+
+The `PositionalParameterExpr` class represents a reference to a parameter
```

Comment:
References to variables in expressions done using DeclRefExpr, which points to the declaration of the variable.  We replace DeclRefExpr with PositionalParameterExpr, so that too needs to be an expression.